### ISSUE TITLE
Support floating point diffs?

### ISSUE THREAD

**Issue by davidedelvento** [paulfitz/daff#59]
I don't know if this request is out of scope for daff, but a really nice feature for a "scientific diff" would be to enable floating point support, perhaps with an appropriate CLI switch or function call argument (off by default, of course).

When enabled, text would be converted to floating point (when possible) and the diff would be performed by floating point difference, in relative terms, say something like d = 2 * (v1 - v2 ) / (v1 + v2). If d <= t for a given threshold t (could be 0.1% be default), the numbers could be considered identical. Also, t could be set to 0, which will be convenient to consider identical things like 1.23e4 and 12300.0

I'll be happy to write the implementation and unit tests for this feature, provided:
- it's of general interest 
- you give me some minimal guidance about how to proceed since I'm not familiar with the development process of daff

**Comment by paulfitz** [comment]
This does sound useful @davidedelvento. Some control over how comparisons are made does seem in scope for a diff tool (e.g. whitespace sensitivity #36).

For development, the main step is to install `haxe`. Make sure you have version 3 or later.  The next step depends what language you are most comfortable in.  If you like javascript, make sure you have `node` installed, and then `make js` will compile daff for javascript and `make ntest_js` will run some tests.

Grepping `ignore_whitespace` in the source code will lead you to all the areas you might need to touch.  Basically three steps involved:
 * setting up and passing around a flag or flags
 * comparing floats for the diff (this is the core bit), see `TableDiff.isEqual`.
 * dealing with floats for indexing - roughly what should happen if a float is in primary key - you could skip this to start with.

**Comment by davidedelvento** [comment]
I'm working on this. The best approach I'm finding implies use of the library polygonal-printf, see what I am doing in https://github.com/davidedelvento/daff/tree/floating_point (not ready for prime time yet, but you should get the idea). Is that acceptable, or do you prefer not adding the external dependency?

**Comment by paulfitz** [comment]
Ok with the dependency if it saves you significant work. I do have to keep my eye on the final size of daff.js for those using it in websites; file size is less important for other uses.

**Comment by davidedelvento** [comment]
I see. My code changes plus this dependency make the new size of the executable 278616 (before it was 245648), or 13% larger. Is that acceptable? I'll be happy to make that dependency optional, if I knew how to do that in haxe :-) 

Anyway, I think I am ready (feature implemented, tests are passing), but on command line it is not behaving as I expected. Do you mind taking a look? (I can make a PR if that makes it easier to look, but it's definitely not ready to be merged)
Maybe I have to do anything with coopy/Index.hx but I am not sure what... Can you please elaborate on the role of this part of the code which isn't very clear to me?

Thanks.

**Comment by paulfitz** [comment]
That's a bit bulky but let's assume the printf dependency is ok.

I looked at your changes and they definitely look along the right lines.

The part of the code that you touched basically controls the *appearance* of the diff, but not which rows are considered equal.  If the problems you are seeing are cases where you believe two rows should be considered the same row, because of float equality, then there is more work to be done for that case.

Equality of rows is decided via a hashing process, since an N-to-N comparison would be too slow for non-trivial tables.  In coopy.Index methods toKey and toKeyByContent, a single row of the table is converted to a key which is used to align rows between tables.  You'll see that whitecase and lowercase options insinuate themselves here. Many possible indexes for different choices of columns included are built up and thrown away or combined during the alignment process.

Before working on figuring out how to muscle floating point stuff in there, let me just ask if this sounds like in fact the issue that is affecting you, or is it something else?

**Comment by davidedelvento** [comment]
Yes, I am seeing cases where I expect two rows to be considered equal, because of float equality (within threshold), and instead they are treated as different. I did the whitespace thingy weaving in toKey*, and I tried to put floating point stuff along similar lines, but it did not work. And I wasn't clear the purpose so it was hard for me to try better. Now that I understand that those two methods are a sort of hashing, I think I can give it another try, however it is different than I initially thought.

In fact, I started from harness/EqualityTest.hx and thought that when my code would be passing the tests that I wrote declaring the desired behavior, I was done (algorithmically) and I was left only with some boilerplate for command line parsing and the likes. Is there any test which is exercising these toKey and toKeyByContent in a way that I can grasp best their expected behavior and add there my additional expectation when floating point threshold is included?

**Comment by paulfitz** [comment]
Hmm, there are javascript tests like `test/test_alignment.js` that check which rows/columns are considered identical (the 'alignment'), which is the handiest way of exercising the indexing code and not the code concerned with diff appearance.  In tests like `harness/BasicTest.hx` you'll see examples of how to get the alignment in haxe, although it isn't actually tested directly there.

You can also exercise indexing using tests like you wrote, but with more rows.  Perhaps force a choice of columns for indexing to simplify things, using `CompareFlags.addPrimaryKey`.  Or have a table with a single column, even simpler.

Stepping back for a second - I find the work you've done already useful, and it fixes an important case for me that doesn't involve float columns contributing to row identity.

Using float columns for identity worries me a bit, because the notion of equality isn't transitive in this case - if A = B and B = C, that doesn't imply that A = C.  I worry about the implications of that for 3-way diffs, which are key to merging, patching, and revision control uses. But maybe my worry is unfounded.

Let's set the index method that exists aside for a moment and go from first principles.  Suppose I give you two tables each with a single column filled with floats.  What algorithm would you propose for deciding which rows in the first table match with which rows in the second table?

**Comment by davidedelvento** [comment]
Hi Paul,
Good that we are having this conversation, because I was only focusing on my own use case, which is pretty straightforward: the two files have the same number of columns and rows, so there is anything to "match" only tell if each "cell" is within the tolerance or now. Certainly for a general purpose tool such as daff I can't make that assumption. So before we start talking algorithms, let's try to focus on the use cases. 

Let's start from the threshold set to zero. In that case, equality is well behaved and we can treat that as you currently do for whitespace and case sensitivity. 

But of course there's more. For the non-zero threshold, my use case is certainly one, which of course is important to me (otherwise I wouldn't be here), but I don't want to make the floating point option restrict to that use case.

Another use case could be comparing, say a table of the distance of the planets from the sun from two different wikipedia pages, that would look like the following

```
Planet name                 distance (km)
Mercury                     57,909,175
Venus                       108,208,930
Earth                       149,597,890
Mars                        227,936,640
Jupiter                     778,412,010
Saturn                      1,426,725,400
Uranus                      2,870,972,200
Neptune                     4,498,252,900
```
but maybe differ in precision and may have different lines (probably a better example would have been a table of densities, which one table reporting some materials and another an overlapping list of materials, but you get the idea).

So I do we want to proceed in this case? In my opinion there are two separate step (which at the moment are unified, if my understanding is correct), namely "matching" lines, and then "diffing" them. For example, if two lines both have "Venus" as planet name, they "match". The value they report may or may not "diff" and that may depend on the threshold being used.

So, how about for the matching part of algorithm we use the one you are currently using, simply ignoring the floating point values that are not equal in the two tables? This should be able to deal with both my use case (which may or may not have text, but has certainly some number that are identical and would "match") and the general case like the one above. 

If we just have table of random numbers, it won't work, but what should we expect out of that?

**Comment by paulfitz** [comment]
That's very helpful @davidedelvento, I understand your use-case now. Agreed that there are two logically separate steps, "matching" and "diffing".  The changes you've already made have taken care of the "diffing" part, but for your use-case the "matching" step also needs updating, as you say.  Your proposal to just ignore columns with floating point values that are not equal may not quite work, since in the matching step we don't yet know about equality.  A revised proposal might be to ignore entire columns if they contain floating point values - that could work, but would leave nothing at all to match on in your case of a table that is all floats.

Here's an idea. Suppose we defined two parameters: 10^-m and 10^-n, where m<n.  When working on matching rows, up until now each row in a table has mapped to a single hash entry, which we can then find quickly in a second table (I know this seems overkill for your case, but it is needed in the general case of insertion/deletion/reordering). Suppose we now relax the constraint, and let a row map to multiple hash entries.  In particular, for a cell containing a float, we will add and subtract 10^-n to it, and then round the two extremes to the nearest 10^-m.  The rounded values may be identical or different. If identical, hashing proceeds exactly as before.  If different, we need to produce two hashes for that row.  Repeating this process across multiple float cells, we may need to double the number of rows multiple times (the odds of this happening depend on choice of m and n).

Matching this way, we can guarantee that cells that differ less than 10^-n will be treated as the same and matched.  The downside is that cells that differ more than 10^-n may well also be treated as the same, since we round off to 10^-m.  But, with a tweak to offset the 10^-m rounding away from aligning with integer boundaries, I think this might prove a scalable solution in practice. 

Backing out of the details - how would you feel about a matching mechanism that guaranteed that:
 * numbers within 0.0001 of each other would be treated as identical
 * numbers differing by more than 0.1 would be treated as distinct
and made no promises about what happened in between?

**Comment by davidedelvento** [comment]
@paulfitz -- sorry for the long silence. I've been caught in other things and this became much less important. In fact, I'm still busy with other things and I don't have time to work further on this, but let me share my thoughts.

I think the `n` and `m` way should work, but would be too complicated for the casual user, whereas a single precision parameter is easier to understand. The question is who is your "average" user? 

Alternatively, we can let the floating point diff work only for my use case (fixed number of rows and columns) which should be straightforward, and then see what other people think of it (and maybe someone else would make a PR extending the functionality).
